app/models/user.rb

```diff
@@ -377,10 +377,14 @@ def retract(target, opts={})
       retraction = Retraction.for(target)
     end
 
-   if target.is_a?(Post)
-     opts[:additional_subscribers] = target.resharers
-     opts[:services] = self.services
-   end
+    if target.is_a?(Post)
+      opts[:additional_subscribers] = target.resharers
+      opts[:services] = self.services
+    elsif target.is_a?(Like)
+      participations.where(target: target.target).first.unparticipate!
+    elsif target.is_a?(Comment)
+      participations.where(target: target.post).first.unparticipate!
+    end
```

Comment:
I think I liked calling it in the callbacks better though. I never checked, passing target like that restricts on type & id?